lib/web/command_utils.go

```diff
@@ -0,0 +1,100 @@
+/*
+
+ Copyright 2023 Gravitational, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+
+*/
+
+package web
+
+import (
+	"encoding/json"
+	"io"
+	"net"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/trace"
+)
+
+// WSConn is a gorilla/websocket minimal interface used by our web implementation.
+type WSConn interface {
+	Close() error
+
+	LocalAddr() net.Addr
+	RemoteAddr() net.Addr
+
+	WriteControl(messageType int, data []byte, deadline time.Time) error
+	WriteMessage(messageType int, data []byte) error
+	NextReader() (messageType int, r io.Reader, err error)
+	ReadMessage() (messageType int, p []byte, err error)
+
+	SetReadDeadline(t time.Time) error
+	PingHandler() func(appData string) error
+	SetPingHandler(h func(appData string) error)
+	PongHandler() func(appData string) error
+	SetPongHandler(h func(appData string) error)
+}
+
+// outEnvelope is an envelope used to wrap messages send back to the client connected over WS.
+type outEnvelope struct {
+	NodeID  string `json:"node_id"`
+	Type    string `json:"type"`
+	Payload []byte `json:"payload"`
+}
+
+type payloadWriter struct {
+	nodeID string
+	// output name, can be stdout, stderr or teleport-error.
+	outputName string
+	// stream is the underlying stream.
+	stream io.Writer
+}
+
+func (p *payloadWriter) Write(b []byte) (n int, err error) {
+	out := &outEnvelope{
+		NodeID:  p.nodeID,
+		Type:    p.outputName,
+		Payload: b,
+	}
+	data, err := json.Marshal(out)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	_, err = p.stream.Write(data)
+	// return the size of the original message as a message send over stream
+	// is larger due to json marshaling and envelope.
+	return len(b), err
+}
+
+func newPayloadWriter(nodeID, outputName string, stream io.Writer) *payloadWriter {
+	return &payloadWriter{
+		nodeID:     nodeID,
+		outputName: outputName,
+		stream:     stream,
+	}
+}
+
+// noopCloserWS is a wrapper around websocket.Conn which does nothing on Close().
+// This struct is used to prevent WS being closed by wrapping stream.
```

Comment:
This feels like a hack. How does the websocket actually get closed when we're done with it if we've no-opped the close method?

Reply:
This code shares a lot of implementation with `TerminalHandler`. The issue here is that `TermianlHandler` closes the WS connection on any error. This implementation allows us to execute a command on multiple servers, so if an error happens on one server, we don't want to close the connection. I think that calling close in multiple places is a bad design decision as `Close()` should be closed once by the "owner" of the connection. Refactoring this code sounds like a risky task and if we want to fix it I think we should do that as a separate task. The no-op `Close()` was the less invasive idea that I had.
I'm open to alternatives.